File: spec/models/world_spec.rb
require "rails_helper"

RSpec.describe World, :type => :model do
  let(:game) { create(:game) }
  let(:world) { game.world }

  before do
    world.build_cells
  end

  let(:cell) { world.cells.find_by(x: 1, y: 1) }
    
  it 'should create a new world world' do
    world.is_a?(World).should be true
  end

  it 'should respond to proper methods' do 
    world.should respond_to(:rows)
    world.should respond_to(:cols)
    world.should respond_to(:cells)
    world.should respond_to(:live_neighbors)
    world.should respond_to(:live_cells)
    world.should respond_to(:randomly_populate)
  end

  it 'should detect a neighbor to the north' do
    north = world.cells.find_by(x: 1, y: 0)
    north.should be_dead
    north.revive!
    north.should be_alive
    world.live_neighbors(cell).count.should eq(1)
  end

  it 'should detect a neighbor to the northeast' do
    northeast = world.cells.find_by(x: 2, y: 0)
    northeast.should be_dead
    northeast.revive!
    northeast.should be_alive
    world.live_neighbors(cell).count.should eq(1)
  end

  it 'should detect a neighbor to the east' do
    east = world.cells.find_by(x: 2, y: 1)
    east.should be_dead
    east.revive!
    east.should be_alive
    world.live_neighbors(cell).count.should eq(1)
  end

  it 'should detect a neighbor to the southeast' do
    southeast = world.cells.find_by(x: 2, y: 2)
    southeast.should be_dead
    southeast.revive!
    southeast.should be_alive
    world.live_neighbors(cell).count.should eq(1)
  end

  it 'should detect a neighbor to the south' do
    south = world.cells.find_by(x: 1, y: 2)
    south.should be_dead
    south.revive!
    south.should be_alive
    world.live_neighbors(cell).count.should eq(1)
  end

  it 'should detect a neighbor to the southwest' do
    southwest = world.cells.find_by(x: 0, y: 2)
    southwest.should be_dead
    southwest.revive!
    southwest.should be_alive
    world.live_neighbors(cell).count.should eq(1)
  end

  it 'should detect a neighbor to the west' do
    west = world.cells.find_by(x: 0, y: 1)
    west.should be_dead
    west.revive!
    west.should be_alive
    world.live_neighbors(cell).count.should eq(1)
  end

  it 'should detect a neighbor to the northwest' do
    northwest = world.cells.find_by(x: 0, y: 0)
    northwest.should be_dead
    northwest.revive!
    northwest.should be_alive
    world.live_neighbors(cell).count.should eq(1)
  end

  it 'should count the number of live cells' do
    world.live_cells.size.should eq(0)
  end

  it 'should randomly populate the world' do
    world.live_cells.size.should eq(0)
    world.randomly_populate
    world.live_cells.size.should_not eq(0)
  end
end

File: spec/javascripts/WorldSpec.js
describe("World", function() {

  beforeEach(function () {
    world = new App.World(3, 3);
  });

  it("should respond to proper methods", function() {
    expect(world.cols).toEqual(3);
    expect(world.rows).toEqual(3);
    expect(world.cells).toEqual(jasmine.any(Array));
    expect(world.cellGrid).toEqual(jasmine.any(Array));
  });

  describe("upon initialization", function() {
    describe("#makeGrid", function () {
      it("should assign 2D array of cell objects to the cellGrid attribute", function () {
        world = new App.World(3, 3);
        expect(world.cellGrid).toEqual(jasmine.any(Array));
        expect(world.cellGrid[0]).toEqual(jasmine.any(Array));
        expect(world.cellGrid.length).toEqual(3);
        expect(world.cells[getRandomInt(0, world.cells.length)]).toEqual(jasmine.any(App.Cell));
        expect(world.cells.length).toEqual(9);
      });
    });
  });

  describe("#liveNeighbors", function() {

    beforeEach(function () {
      cell = world.cellGrid[1][1];
      expect(world.liveNeighbors(cell).length).toEqual(0);
    });

    it("should detect a neighbor to the north", function () {
      world.cellGrid[0][1].revive();
      expect(world.liveNeighbors(cell).length).toEqual(1);
    });

    it("should detect a neighbor to the northeast", function () {
      world.cellGrid[0][2].revive();
      expect(world.liveNeighbors(cell).length).toEqual(1);
    });

    it("should detect a neighbor to the east", function () {
      world.cellGrid[1][2].revive();
      expect(world.liveNeighbors(cell).length).toEqual(1);
    });

    it("should detect a neighbor to the southeast", function () {
      world.cellGrid[2][2].revive();
      expect(world.liveNeighbors(cell).length).toEqual(1);
    });

    it("should detect a neighbor to the south", function () {
      world.cellGrid[2][1].revive();
      expect(world.liveNeighbors(cell).length).toEqual(1);
    });

    it("should detect a neighbor to the southwest", function () {
      world.cellGrid[2][0].revive();
      expect(world.liveNeighbors(cell).length).toEqual(1);
    });

    it("should detect a neighbor to the west", function () {
      world.cellGrid[1][0].revive();
      expect(world.liveNeighbors(cell).length).toEqual(1);
    });

    it("should detect a neighbor to the northwest", function () {
      world.cellGrid[0][0].revive();
      expect(world.liveNeighbors(cell).length).toEqual(1);
    });
  });

  describe("counting and killing cells", function() {
    
    describe("#liveCells", function() {
      
      it("returns an array of all live cells", function () {
        expect(world.liveCells().length).toEqual(0);
        world.cellGrid[1][1].revive();
        world.cellGrid[2][2].revive();
        expect(world.liveCells().length).toEqual(2);
        world.cellGrid[2][2].die();
        expect(world.liveCells().length).toEqual(1);

        describe("#killAllCells", function() {
          
          it("should set alive to false for all cells", function () {
            expect(world.liveCells().length).toEqual(1);
            world.killAllCells();
            expect(world.liveCells().length).toEqual(0);
          });
        });
      });
    });
  });
});